rfcs/20231129-tuning-api/README.md

```diff
@@ -0,0 +1,134 @@
+# RFC for Auto-Tuning API
+
+## Introduction & Motivation
+
+Auto-tuning is a feature supported as a mode in PyTorch's [torch.compile](https://pytorch.org/docs/stable/generated/torch.compile.html) function 
+and in TensorFlow via [XLA](https://github.com/sourcecode369/tensorflow-1/blob/9f446aba8aaeb2b3c4c6e5ba1ab4cf31494b8a64/tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc#L279).
+While the median model level improvement is generally modest, there are cases where there are 
+large speedups. This has been observed on Intel hardware (see [Using tuning to evaluate the level of oneDNN performance](https://github.com/intel-innersource/libraries.performance.math.onednn/pull/5931)) 
+and externally on other [hardware](https://mangpo.net/papers/xla-autotuning-pact2021.pdf).
+
+The goal is to allow end users to optionally try auto-tuning in the case where out of the
+box performance is insufficient.
+
+## Proposal
+OneDNN should implement auto tuning as a feature that can be exposed to end users of frameworks.
+Major requirements for framework integration:
+1) No changes to primitive API.
+2) Primitive and kernel cache states should not be affected.
+3) No regressions after tuning.
+4) Simple knob to enable/disable tuning.
+
+### Option 1- (Recommended)
+Basic Usage Scenario
+```c
+create_primitive(prim); 
+dnnl_set_tune(true);   
+execute(prim);  //tuning happens here
+dnnl_set_tune(false); 
```

Comment:
Another potential "creation" scenario I've seen recently: a user wants to probe the primitive cache, that is skip creation (oneDNN returns an error) in case of a primitive cache miss. All this (together with tuning) could be combined under "create mode" sub-attribute.

Reply:
> Can you give an example of a conflict?

The core issue is that there is a significant internal state (visible via realized performance) based on execution order. This can cause reproduction issues depending the execution order. The ordering issues that come to my mind are
```C++
create(p);
tune_create(p); // Primitive cache hit, get default version
```
```C++
tune_create(p); // Get tuned version
create(p); // Primitive cache hit, get tuned version
... // create many primitives
create(p); // Primitive cache miss, get default version
```
This becomes significantly more complicated in a multi-threaded scenario where the ordering can vary from run to run.

Reply:
The config iteration number is now part of the `pd_key`. Therefore any create that does not set the that iteration value will go to the default version. That version won't change until updated in the "finalize" part of tuning which will update the default key to point to the tuned implementation. In the case of a cache miss on the default version, the lookup table  would be used to create the optimized implementation.

To me it seems like primitive cache management has to be addressed in a similar way regardless of the api.